### PR TITLE
Plans 2023: Hide custom domain feature if domain credit has been redeemed

### PR DESCRIPTION
**Pull request by jeyip**
Related to https://github.com/Automattic/martech/issues/2850 and D145279-code

## Proposed Changes

Currently, our pricing grid still shows "Free domain for one year" in the logged-in /plans on a paid site even if its domain credits _have already been used_.

There are users who have mistaken this to mean that they could gain another free domain for a year upon upgrading their plan. Because of this, we should exclude the "Free domain for one year" feature if one has already been redeemed for their site.

### Before
Screenshot 2024-04-15 at 1 24 52 PM

Screenshot 2024-04-15 at 1 52 14 PM

### After
Screenshot 2024-04-15 at 1 24 44 PM

Screenshot 2024-04-15 at 1 52 31 PM

## Testing Instructions

- Test alongside D145279-code
  - Sandbox public-api.wordpress.com 
  - Patch D145279
- Check out this branch
- Set up a new site with a free plan 
- Visit calypso.localhost:3000/plans/{site_slug_for_new_site}
- Verify that we still display the "Free domain for one year" feature in the features grid and the comparison grid
- Upgrade to a plan that provides free domain for a year ( starter, explorer, creator, entrepreneur )
- Revisit calypso.localhost:3000/plans/{site_slug_for_new_site}
- Verify that we no longer display the "Free domain for one year" for plan upgrades or downgrades

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/state/sites/plans/assembler.js [hunk 1]
```diff
@@ -17,6 +17,7 @@ export const createSitePlanObject = ( plan ) => {
 		formattedPrice: plan.formatted_price,
 		freeTrial: Boolean( plan.free_trial ),
 		hasDomainCredit: Boolean( plan.has_domain_credit ),
+		hasRedeemedDomainCredit: Boolean( plan.has_redeemed_domain_credit ),
```
Comment (chriskmnds) [inline client/state/sites/plans/assembler.js]: We probably shouldn't be extending this with more properties. The plan is to deprecate this state slice completely in favor of the plans data-store (see https://github.com/Automattic/wp-calypso/issues/86638). I would consider adding this to the return value from [use-site-plans](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/plans/queries/use-site-plans.ts) and use `useCurrentPlan` hook from there to bring it into `plans-features-main`.

Comment (jeyip) [reply client/state/sites/plans/assembler.js]: Thanks for the feedback Christos -- exploring this now 🕵️

Comment (jeyip) [reply client/state/sites/plans/assembler.js]: Feedback should be addressed by https://github.com/Automattic/wp-calypso/pull/89060/commits/0990b75fd673e06fb0daa9e5a42553cb86ee3c28 @chriskmnds

File: packages/calypso-products/src/main.ts [hunk 2]
```diff
@@ -853,7 +853,7 @@ export const chooseDefaultCustomerType = ( {
 }: {
 	currentCustomerType: string;
 	selectedPlan?: string;
-	currentPlan: { productSlug: PlanSlug };
+	currentPlan: { productSlug: PlanSlug | undefined };
```
Comment (jeyip) [inline packages/calypso-products/src/main.ts:856]: **Note:**
The `chooseDefaultCustomerType` was always being passed `undefined` values for `currentPlan`. Further below, there's actually a check for a falsey `currentPlan` value. I'm unsure why typescript wasn't throwing an error before, but I've explicitly added the `undefined` type here.

**Comment by jeyip** [comment]
Planning to proceed with this approach tomorrow and merge as long as there's no other feedback 🙂

**Pull request merged**